src/ExecClient.cs

```diff
@@ -0,0 +1,116 @@
+using System;
+using System.Buffers;
+using System.Net.WebSockets;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace k8s
+{
+    public class ExecClient : IDisposable
+    {
+        private readonly ClientWebSocket socket;
+        private readonly CancellationTokenSource cts;
+        private Task runLoop;
+
+        public ExecClient(ClientWebSocket socket)
+        {
+            this.socket = socket ?? throw new ArgumentNullException(nameof(socket));
+            this.cts = new CancellationTokenSource();
+        }
+
+        public event EventHandler ConnectionClosed;
+        public event EventHandler<string> StandardOutputReceived;
+        public event EventHandler<string> StandardErrorReceived;
+
+        public void Start()
+        {
+            if (this.runLoop == null)
+            {
+                this.runLoop = this.RunLoop(this.cts.Token);
+            }
+        }
+
+        public void Stop(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (this.runLoop != null)
+            {
+                this.cts.Cancel();
+                this.runLoop.Wait(cancellationToken);
+            }
+        }
+
+        public void Dispose()
+        {
+            this.Stop();
+        }
+
+        public async Task Write(string command, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            // Send:
+            //  - [uint32: stream index = 0 for stdin]
+            //  - [byte[]: command as UTF8 value]
+            var commandLength = Encoding.UTF8.GetByteCount(command);
+            var buffer = ArrayPool<byte>.Shared.Rent(commandLength + 4);
+
+            try
+            {
+                // The first 4 bytes represent the stream index. For stdin this is 0
+                buffer[0] = 0;
+                buffer[1] = 0;
+                buffer[2] = 0;
+                buffer[3] = 0;
+
+                // Copy the command to the buffer, starting at offset 4
+                Encoding.UTF8.GetBytes(command, 0, command.Length, buffer, 4);
+
+                ArraySegment<byte> segment = new ArraySegment<byte>(buffer, 0, commandLength + 4);
+                await this.socket.SendAsync(segment, WebSocketMessageType.Binary, true, cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                ArrayPool<byte>.Shared.Return(buffer);
+            }
+        }
+
+        protected async Task RunLoop(CancellationToken cancellationToken)
+        {
+            // Get a 1KB buffer
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(1024 * 1024);
+            var segment = new ArraySegment<byte>(buffer);
+
+            while (!cancellationToken.IsCancellationRequested && this.socket.CloseStatus == null)
+            {
+                // We always get data in this format:
```

Comment:
Ultimately we're going to want to pull this logic out so that we can use it for Attach and PortForward, but we can do that in a different PR...

Reply:
Yeah, I didn't have the time to look at the protocol for Attach yet (not sure whether it uses stream multiplexing or not).

I had a cursory look at `PortForward` and I believe it's simpler - there is no multiplexing so you can just use `ReceiveAsync`/`SendAsync` to talk with the remote socket.
You may want to add code which wraps the `ClientWebSocket` in `Stream`-like class, though.

That said, one thing at a time 😄